Add functional methods for disks and Disk class

mziccard opened the pull request:
This PR also includes integration tests for disks images and snapshots.

mziccard commented on gcloud-java-compute/src/test/java/com/google/gcloud/compute/it/ITComputeTest.java:
```diff
@@ -855,4 +875,464 @@ public void testListGlobalAddresses() throws InterruptedException {
+  }
+
+  @Test
+  public void testListDisksAndSnapshots() throws InterruptedException {
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [The method testListDisksAndSnapshots() has an NPath complexity of 256](https://www.codacy.com/app/mziccard/gcloud-java/file/2372551756/issues/source?bid=3140968&fileBranchId=3223208#l1066)

mziccard commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java:
```diff
@@ -918,6 +918,54 @@ public static ImageFilter notEquals(ImageField field, long value) {
+     *
+     * @see <a href="https://github.com/google/re2/wiki/Syntax">RE2</a>
+     */
+    public static DiskFilter equals(DiskField field, String value) {
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [The method name and parameter number are suspiciously close to equals(Object)](https://www.codacy.com/app/mziccard/gcloud-java/file/2372551672/issues/source?bid=3140968&fileBranchId=3223208#l938)

mziccard replied:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [The method name and parameter number are suspiciously close to equals(Object)](https://www.codacy.com/app/mziccard/gcloud-java/file/2383968021/issues/source?bid=3140968&fileBranchId=3223208#l938)

mziccard commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java:
```diff
@@ -918,6 +918,54 @@ public static ImageFilter notEquals(ImageField field, long value) {
+    /**
+     * Returns an equals filter for the given field and long value.
+     */
+    public static DiskFilter equals(DiskField field, long value) {
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [The method name and parameter number are suspiciously close to equals(Object)](https://www.codacy.com/app/mziccard/gcloud-java/file/2372551672/issues/source?bid=3140968&fileBranchId=3223208#l956)

mziccard replied:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [The method name and parameter number are suspiciously close to equals(Object)](https://www.codacy.com/app/mziccard/gcloud-java/file/2383968021/issues/source?bid=3140968&fileBranchId=3223208#l956)

mziccard commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java:
```diff
@@ -1161,6 +1201,124 @@ public Operation deprecate(ImageId image,
+    return listDisks(zone, options(), optionMap(options));
+  }
+
+  private static Page<Disk> listDisks(final String zone, final ComputeOptions serviceOptions,
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused private methods such as 'listDisks(String,ComputeOptions,Map)'.](https://www.codacy.com/app/mziccard/gcloud-java/file/2372551675/issues/source?bid=3140968&fileBranchId=3223208#l1254)

mziccard commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java:
```diff
@@ -1161,6 +1201,124 @@ public Operation deprecate(ImageId image,
+    return listDisks(options(), optionMap(options));
+  }
+
+  private static Page<Disk> listDisks(final ComputeOptions serviceOptions,
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused private methods such as 'listDisks(ComputeOptions,Map)'.](https://www.codacy.com/app/mziccard/gcloud-java/file/2372551675/issues/source?bid=3140968&fileBranchId=3223208#l1282)

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Disk.java:
```diff
@@ -0,0 +1,249 @@
+   *     the disk was not found
+   * @throws ComputeException upon failure
+   */
+  public Operation createSnapshot(String snapshot, OperationOption... options) {
```
I think I missed that in the compute.create(SnapshotInfo). Any reason to change the pattern and return `null` in create in this case?

mziccard replied:
Also @ajkannan had noticed this before. The reason is that `createSnapshot` is an API of disk resources. Usually to create a resource we use the following endpoint (the example shows a disk):
```
POST /projects/<project>/zones/<zone>/disks/<disk>/insert
```
To create a snapshot we don't have an `/insert` endpoint, instead the endpoint is:
```
POST /projects/<project>/zones/<zone>/disks/<disk>/createSnapshot
```
Usually requests to `/insert` never return 404, `/createSnapshot` instead returns 404 if the disk does not exist. If you prefer for consistency we can avoid catching the 404 in this case.

aozarov replied:
What would happen if zone not found in either of your examples?

aozarov commented:
Looks good. I only have one question. @ajkannan do you mind to give it another pass and especially look at the tests/docs (which I completely ignored). If not, I will probably have time for a more thorough review tomorrow afternoon.

ajkannan commented:
Yes, I can give this a pass now.

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java:
```diff
@@ -1586,6 +1634,110 @@ public static ImageListOption fields(ImageField... fields) {
+
+    /**
+     * Returns an option to specify the disk's fields to be returned by the RPC call. If this option
+     * is not provided, all disk's fields are returned. {@code DiskOption.fields} can be used to
```
nit: s/all disk's fields/all disk fields/

mziccard replied:
The saxon genitive is used for all `field` selector added so far... do you feel strongly about this?:)

ajkannan replied:
It can stay :)

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java:
```diff
@@ -1586,6 +1634,110 @@ public static ImageListOption fields(ImageField... fields) {
+     * Returns an option to specify the disk's fields to be returned by the RPC call. If this option
+     * is not provided, all disk's fields are returned. {@code DiskOption.fields} can be used to
+     * specify only the fields of interest. {@link Disk#diskId()},
+     * {@link DiskConfiguration#diskType()} and {@link SnapshotDiskConfiguration#sourceSnapshot()}
```
Nit: could add "either" after "and" to resolve a tiny bit of grammatical ambiguity.

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java:
```diff
@@ -1586,6 +1634,110 @@ public static ImageListOption fields(ImageField... fields) {
+
+    /**
+     * Returns an option to specify the disk's fields to be returned by the RPC call. If this option
+     * is not provided, all disk's fields are returned. {@code DiskListOption.fields} can be used to
```
Could apply same nits as above (disk's --> disk, add "either" after "and" in second to last line)

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java:
```diff
@@ -1161,6 +1201,124 @@ public Operation deprecate(ImageId image,
+    }
+  }
+
+  private static Function<com.google.api.services.compute.model.Disk, Disk> diskFromPb(
```
Just curious, any reason to put this method here rather than having a package-scope static `Function` in the Disk class (like in Storage's Blob?).  If a user is using a lot of paging and list calls (perhaps an uncommon use case), could it be slightly more efficient to have the `Function` declared once statically in Disk rather than returning a new `Function` each time?  I don't feel strongly about changing it, I'm fine with keeping it as is if you prefer that.

mziccard replied:
The function in `Blob` takes a tuple and is only used in `BatchResponse submit(BatchRequest batchRequest)` to allow `transformBatchResult` to be more generic. The function takes a tuple, if we have a collection I think it will take more (both time and memory) to create a tuple `<Strorage, Disk>` for each item in the collection and then call the `Function` rather than having a factory method for the transformation function which we only need to call once for each collection transformation.

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Disk.java:
```diff
@@ -0,0 +1,249 @@
+    Builder(Compute compute, DiskId diskId, DiskConfiguration diskConfiguration) {
+      this.compute = compute;
+      this.infoBuilder = new DiskInfo.BuilderImpl(diskId, diskConfiguration);
+      this.infoBuilder.diskId(diskId);
```
Is this line and the line below unnecessary because `diskId` and `diskConfiguration` are constructor arguments?

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Disk.java:
```diff
@@ -0,0 +1,249 @@
+  }
+
+  /**
+   * Returns the disk's {@code Compute} object used to issue requests.
```
nit: If you want, you could do a `@link` here

mziccard replied:
I didn't use `@link` because Compute should be already linked in the method's javadoc as the return type.

ajkannan commented:
Question: is [disk resizing](https://cloud.google.com/compute/docs/reference/latest/disks/resize) something that we don't want to implement right now?

ajkannan commented on gcloud-java-compute/src/test/java/com/google/gcloud/compute/DiskTest.java:
```diff
@@ -0,0 +1,452 @@
+  private static final SnapshotId SNAPSHOT = SnapshotId.of("project", "snapshot");
+  private static final ImageId IMAGE = ImageId.of("project", "image");
+  private static final String SNAPSHOT_ID = "snapshotId";
+  private static final String IMAGE_ID = "snapshotId";
```
s/snapshotId/imageId/

ajkannan commented on gcloud-java-compute/src/test/java/com/google/gcloud/compute/it/ITComputeTest.java:
```diff
@@ -855,4 +875,464 @@ public void testListGlobalAddresses() throws InterruptedException {
+    DiskInfo diskInfo =
+        DiskInfo.of(diskId, StandardDiskConfiguration.of(DiskTypeId.of(ZONE, "pd-ssd"), 100L));
+    Operation operation = compute.create(diskInfo);
+    while (!operation.isDone()) {
```
This might be a silly question, but does the Operation status "done" correspond to the Disk status "ready"?

mziccard replied:
Yes the operation is done when the disk is READY and yes it is returned in CREATING status.

ajkannan commented on gcloud-java-compute/src/test/java/com/google/gcloud/compute/it/ITComputeTest.java:
```diff
@@ -855,4 +875,464 @@ public void testListGlobalAddresses() throws InterruptedException {
+    assertNotNull(remoteDisk.<ImageDiskConfiguration>configuration().sourceImageId());
+    assertEquals(DiskConfiguration.Type.IMAGE, remoteDisk.configuration().type());
+    assertNotNull(remoteDisk.configuration().sizeGb());
+    assertEquals("pd-standard", remoteDisk.configuration().diskType().diskType());
```
I think makes me wonder a little about diskType naming.  Could we make the string disk type inside DiskTypeId just "type"?

mziccard replied:
I don't mind doing this change but I would leave it for a different PR.

ajkannan replied:
Sure, sounds good.  Maybe we can create an issue for it?

mziccard replied:
Opened #887

ajkannan commented on gcloud-java-compute/src/test/java/com/google/gcloud/compute/it/ITComputeTest.java:
```diff
@@ -855,4 +875,464 @@ public void testListGlobalAddresses() throws InterruptedException {
+    assertEquals(DiskInfo.CreationStatus.READY, remoteDisk.creationStatus());
+    assertNotNull(remoteDisk.creationTimestamp());
+    assertNotNull(remoteDisk.id());
+    assertTrue(remoteDisk.configuration() instanceof ImageDiskConfiguration);
```
Nit: if you want, you could save some configuration() calls and casting if you save remoteDisk.configuration() to a variable (here and for similar cases below).

ajkannan commented on gcloud-java-compute/src/test/java/com/google/gcloud/compute/it/ITComputeTest.java:
```diff
@@ -855,4 +875,464 @@ public void testListGlobalAddresses() throws InterruptedException {
+    assertNull(snapshot.sourceDiskId());
+    assertNull(snapshot.storageBytes());
+    assertNull(snapshot.storageBytesStatus());
+    SnapshotInfo.of(SnapshotId.of(snapshotName), snapshotDiskId);
```
What's this line's purpose?

ajkannan commented on gcloud-java-compute/src/test/java/com/google/gcloud/compute/it/ITComputeTest.java:
```diff
@@ -855,4 +875,464 @@ public void testListGlobalAddresses() throws InterruptedException {
+    Page<Image> imagePage = compute.listImages(IMAGE_PROJECT, Compute.ImageListOption.filter(
+        Compute.ImageFilter.equals(Compute.ImageField.ARCHIVE_SIZE_BYTES, 365056004L)));
+    Iterator<Image> imageIterator = imagePage.iterateAll();
+    while (imageIterator.hasNext()) {
```
Do you think that we should assert that we got at least one result?  Suppose our filter construction was wrong, we could get no results, and this test would pass.

ajkannan commented on gcloud-java-compute/src/test/java/com/google/gcloud/compute/it/ITComputeTest.java:
```diff
@@ -855,4 +875,464 @@ public void testListGlobalAddresses() throws InterruptedException {
+    Page<Image> imagePage =
+        compute.listImages(IMAGE_PROJECT, Compute.ImageListOption.fields(Compute.ImageField.ID));
+    Iterator<Image> imageIterator = imagePage.iterateAll();
+    while (imageIterator.hasNext()) {
```
Do you think we should assert that this while loop is entered at least once?

ajkannan commented on gcloud-java-compute/src/test/java/com/google/gcloud/compute/it/ITComputeTest.java:
```diff
@@ -855,4 +875,464 @@ public void testListGlobalAddresses() throws InterruptedException {
+      count++;
+    }
+    assertEquals(2, count);
+    compute.delete(firstDiskId);
```
Here and in the method below you don't wait for the operation to complete before exiting the test, unlike in the methods above.  Any particular reason for that?

mziccard replied:
I generally don't wait when the method is not testing `delete` and there are no constraint on the maximum number of that resources that can be created. This allows the tests to run faster.

ajkannan commented on gcloud-java-compute/src/test/java/com/google/gcloud/compute/it/ITComputeTest.java:
```diff
@@ -855,4 +875,464 @@ public void testListGlobalAddresses() throws InterruptedException {
+    }
+    Disk remoteDisk = compute.get(diskId);
+    ImageInfo imageInfo = ImageInfo.of(imageId, DiskImageConfiguration.of(diskId));
+    operation = compute.create(imageInfo);
```
Did you also consider creating an Image using a StorageImageConfiguration?

mziccard replied:
Yes but it's not that easy. You need a raw image to be loaded in cloud storage. Images are quite big and we can not load them on the go in our ITs. Having the raw image loaded in one of our buckets makes the ITs non-repeatable by users. Given this I decided to skip that test.

ajkannan commented:
Done with my pass.

mziccard commented:
I fixed all comments. To add resize support I updated the compute dependency which also removed `MaintenanceWindow`'s list from `Zone`. Fixed that as well.

mziccard commented on gcloud-java-compute/src/test/java/com/google/gcloud/compute/it/ITComputeTest.java:
```diff
@@ -855,4 +874,486 @@ public void testListGlobalAddresses() throws InterruptedException {
+  }
+
+  @Test
+  public void testListDisksAndSnapshots() throws InterruptedException {
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [The method testListDisksAndSnapshots() has an NPath complexity of 256](https://www.codacy.com/app/mziccard/gcloud-java/file/2383968105/issues/source?bid=3140968&fileBranchId=3223208#l1073)

mziccard commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java:
```diff
@@ -1161,6 +1201,141 @@ public Operation deprecate(ImageId image,
+    return listDisks(zone, options(), optionMap(options));
+  }
+
+  private static Page<Disk> listDisks(final String zone, final ComputeOptions serviceOptions,
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused private methods such as 'listDisks(String,ComputeOptions,Map)'.](https://www.codacy.com/app/mziccard/gcloud-java/file/2383968024/issues/source?bid=3140968&fileBranchId=3223208#l1254)

mziccard commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java:
```diff
@@ -1161,6 +1201,141 @@ public Operation deprecate(ImageId image,
+    return listDisks(options(), optionMap(options));
+  }
+
+  private static Page<Disk> listDisks(final ComputeOptions serviceOptions,
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused private methods such as 'listDisks(ComputeOptions,Map)'.](https://www.codacy.com/app/mziccard/gcloud-java/file/2383968024/issues/source?bid=3140968&fileBranchId=3223208#l1282)

ajkannan commented:
`MAINTENANCE_WINDOW` still exists in the `ZoneField` enum (in `Compute.java`) -- is that intentional or should it be removed?

mziccard commented:
> MAINTENANCE_WINDOW still exists in the ZoneField enum (in Compute.java) -- is that intentional or should it be removed?

I forgot, it should be removed.

ajkannan commented:
LGTM, feel free to merge after removing the field from the ZoneFields.
